### ISSUE TITLE
`Matrix.echelon_form()` should not ignore `transformation=True` option

### ISSUE THREAD

**Issue by maxale** [sagemath/sage#37480]
### Steps To Reproduce

[Documentation for `.echelon_form()`](https://doc.sagemath.org/html/en/reference/matrices/sage/matrix/matrix2.html#sage.matrix.matrix2.Matrix.echelon_form) says:

> `transformation` – boolean. Whether to also return the transformation matrix. Some matrix backends do not provide this information, in which case this option is ignored.

The example below illustrates how such ignoring happens in practice:

```
sage: M = Matrix([[1,2,3],[4,5,6],[7,8,9]])
sage: M.echelon_form(transformation = True)
(
[1 2 3]  [ 0  2 -1]
[0 3 6]  [ 0  7 -4]
[0 0 0], [ 1 -2  1]
)
sage: M.change_ring(QQ).echelon_form(transformation = True)
[ 1  0 -1]
[ 0  1  2]
[ 0  0  0]
```

### Expected Behavior

It is a bad design decision to ignore given options. Instead, if something is not supported, Sage should produce an error. This will inform the user about the issue and encourage to look for a workaround (e.g. using a different backend). If the matrix backend does not provide a transformation, the error message should clearly communicate that to the user, including the information about what backend was used.

### Actual Behavior

Sage silently ignores `transformation=True` parameter in some cases, without giving a reason. This makes the output not well determined as it can arbitrarily return one or two matrices, and hides the reason for missing transformation from the user.

### Environment

```markdown
- **OS**: Ubuntu 22.04.4 LTS
- **Sage Version**: 10.3.beta7
```

### Checklist

- [X] I have searched the existing issues for a bug report that matches the one I want to file, without success.
- [X] I have read the documentation and troubleshoot guide

**Comment by HenryWu2101** [comment]
I think this is due to the fact that the `echelon_form` method in `matrix_rational_dense ` does not support the transformation argument, and this is true for all other matrix classes that have their own `echelon_form meth::`. So when lifted to QQ, it ignores the argument. A discussion of different behaviors of `echelon_form` in different matrix class is seen here also #39917

**Comment by HenryWu2101** [comment]
This is also mentioned in a previous issue #29177

**Comment by vneiger** [comment]
For good reasons about efficiency, some libraries are not computing the transformation when only seeking an echelon form. If one really wants a transformation, they could rely on decompositions such as PLU (but Sage is currently not always / not often calling fast implementations for such decompositions). So I agree that there is no easy fix and would propose to close this issue.

**Comment by maxale** [comment]
I do not quite follow how the assertion that some libraries do not compute transformation affects the current issue. Please note that this issue is about the Sage interface of `Matrix.echelon_form()`, not more not less. It's fine for a library to not support computing the transformation, however, if a user requests a transformation via `transformation=True` parameter but it was not computed (for whatever reason), Sage should not return a partial result but rather raise an exception. The user then will decide whether it's ok to not have a transformation and drop the `transformation` parameter (or set it to `False`), or to look for a workaround for getting the transformation computed.

Also, I do not buy the _"no easy fix"_ as a reason for closing the issue. If the issue exists, it should not be closed on the ground that it may be hard to be fix. Someone may be willing to do the "hard work" and fix the issue later on.

**Comment by HenryWu2101** [comment]
> if a user requests a transformation via transformation=True parameter but it was not computed (for whatever reason), Sage should not return a partial result but rather raise an exception. The user then will decide whether it's ok to not have a transformation and drop the transformation parameter (or set it to False), or to look for a workaround for getting the transformation computed.

In this case, may I propose that the best way to resolve this issue at the moment would be to raise exception for all matrix class that use their own `echelon_form meth` and suggest the user to find the transformation matrix using the result (by `solve` directly or whatever). I think **vneiger** also had a point about how some cases it is not ideal to implement this feature inside `echelon_form`, I would agree with him from my understanding of the current framework.

**Comment by vneiger** [comment]
> In this case, may I propose that the best way to resolve this issue at the moment would be to raise exception for all matrix class that use their own `echelon_form meth` and suggest the user to find the transformation matrix using the result (by `solve` directly or whatever).

Yes, raising an exception for classes that do not support this option would be fine, I suppose. It is not very satisfactory for other reasons I mention below, but we could see it as an "easy fix" for this specific issue, if we ignore the more global picture (and that might be preferable temporarily, to avoid mixing different issues together, which is what led me to wrongly suggest to close this issue).

> I think **vneiger** also had a point about how some cases it is not ideal to implement this feature inside `echelon_form`, I would agree with him from my understanding of the current framework.

We could make this method always accept the option by adding a small piece of code that computes the transformation, as you say using some `solve`. I did not want to propose this initially, by fear of it potentially ruining efficiency in some cases. I'm not really sure what is best between this and the exception. Maybe a very weak preference for the exception.

Something that is troubling me more globally is that "echelon form with transformation" is basically just one more flavor of the classical LU-like decompositions. Sage already has a (slow, general) LU method for PLU decomposition, for some types of matrices it has a PLE or PLUQ decomposition but that one is not used to make specialize the LU method into something fast, for some types it has echelon forms with transformations and for some others not... it would make sense to make all of this a bit more uniform.

**Comment by HenryWu2101** [comment]
I went ahead and took the exception path anyways. I think we should first merge that to develop, then keep this issue for future fix
